### ISSUE TITLE
Intermittent failures in interface_usdt_mempool.py

### ISSUE THREAD

**Issue by dergoegge** [bitcoin/bitcoin#27380]
https://cirrus-ci.com/task/5779522121891840

```
[0;34m node0 2023-03-31T11:45:58.812384Z (mocktime: 2023-04-14T11:46:01Z) [http] [httpserver.cpp:257] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:38238 [0m
[0;34m node0 2023-03-31T11:45:58.812543Z (mocktime: 2023-04-14T11:46:01Z) [httpworker.2] [rpc/request.cpp:179] [parse] [rpc] ThreadRPCServer method=getpeerinfo user=__cookie__ [0m
[0;36m test  2023-03-31T11:45:58.814000Z TestFramework (INFO): Hooking into mempool:rejected tracepoint... [0m
[0;36m test  2023-03-31T11:45:59.798000Z TestFramework (INFO): Sending invalid transaction... [0m
[0;36m test  2023-03-31T11:45:59.799000Z TestFramework.p2p (DEBUG): Send message to 127.0.0.1:15944: msg_tx(tx=CTransaction(nVersion=2 vin=[CTxIn(prevout=COutPoint(hash=76b2ca9a8c3ffa3c454db9591dfaf5c4979926154e33a5bef67412ea0eaa16c5 n=0) scriptSig= nSequence=0)] vout=[CTxOut(nValue=49.99968800 scriptPubKey=51202913b252fe537830f843bfdc5fa7d20ba48639a87c86ff837b92d083c55ad7c1)] wit=CTxWitness(CScriptWitness(51,c00000000000000000000000000000000000000000000000000000000000000001)) nLockTime=0)) [0m
[0;36m test  2023-03-31T11:45:59.799000Z TestFramework.p2p (DEBUG): Send message to 127.0.0.1:15944: msg_ping(nonce=00000002) [0m
[0;34m node0 2023-03-31T11:45:59.799697Z (mocktime: 2023-04-14T11:46:01Z) [msghand] [net_processing.cpp:3169] [ProcessMessage] [net] received: tx (133 bytes) peer=0 [0m
[0;34m node0 2023-03-31T11:45:59.800472Z (mocktime: 2023-04-14T11:46:01Z) [msghand] [txmempool.cpp:658] [check] [mempool] Checking mempool with 2 transactions and 2 inputs [0m
[0;34m node0 2023-03-31T11:45:59.800746Z (mocktime: 2023-04-14T11:46:01Z) [msghand] [net_processing.cpp:4174] [ProcessMessage] [mempoolrej] c917a71e43b825756c1d45985fcc87f8c245db0654f5eeb000a92dc2cccb1d6c from peer=0 was not accepted: min relay fee not met, 0 < 104 [0m
[0;34m node0 2023-03-31T11:45:59.800913Z (mocktime: 2023-04-14T11:46:01Z) [msghand] [net_processing.cpp:3169] [ProcessMessage] [net] received: ping (8 bytes) peer=0 [0m
[0;34m node0 2023-03-31T11:45:59.800953Z (mocktime: 2023-04-14T11:46:01Z) [msghand] [net.cpp:2841] [PushMessage] [net] sending pong (8 bytes) peer=0 [0m
[0;36m test  2023-03-31T11:45:59.801000Z TestFramework.p2p (DEBUG): Received message from 127.0.0.1:15944: msg_pong(nonce=00000002) [0m
[0;34m node0 2023-03-31T11:45:59.849969Z (mocktime: 2023-04-14T11:46:01Z) [http] [httpserver.cpp:257] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:38238 [0m
[0;36m test  2023-03-31T11:45:59.850000Z TestFramework (INFO): Polling buffer... [0m
[0;34m node0 2023-03-31T11:45:59.850174Z (mocktime: 2023-04-14T11:46:01Z) [httpworker.3] [rpc/request.cpp:179] [parse] [rpc] ThreadRPCServer method=getrawmempool user=__cookie__ [0m
[0;36m test  2023-03-31T11:45:59.898000Z TestFramework (INFO): Ensuring mempool:rejected event was handled successfully... [0m
[0;36m test  2023-03-31T11:45:59.898000Z TestFramework (ERROR): Assertion failed [0m
[0;36m                                   Traceback (most recent call last):[0m
[0;36m                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 132, in main[0m
[0;36m                                       self.run_test()[0m
[0;36m                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 339, in run_test[0m
[0;36m                                       self.rejected_test()[0m
[0;36m                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 323, in rejected_test[0m
[0;36m                                       assert_equal(EXPECTED_REJECTED_EVENTS, handled_rejected_events)[0m
[0;36m                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 56, in assert_equal[0m
[0;36m                                       raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))[0m
[0;36m                                   AssertionError: not(1 == 0)[0m
[0;36m test  2023-03-31T11:45:59.899000Z TestFramework (DEBUG): Closing down network thread [0m
[0;36m test  2023-03-31T11:45:59.949000Z TestFramework (INFO): Stopping nodes [0m
[0;36m test  2023-03-31T11:45:59.950000Z TestFramework.node0 (DEBUG): Stopping node [0m
```

**Comment by dergoegge** [comment]
Have seen this a couple times by now.

cc @0xB10C @virtu

**Comment by 0xB10C** [comment]
Had a brief look. The `mempoolrej` log message shows that we indeed rejected the transaction but the test didn't see the reject. 

Can't reproduce this locally with about 20 test runs. All pass. Might be a CI timing problem then? Will try playing around with the `bpf.perf_buffer_poll()` timeout.

Will investigate further.

**Comment by 0xB10C** [comment]
Other CI runs with similar behavior:

- https://cirrus-ci.com/task/6650565413830656?logs=ci#L3531

**Comment by 0xB10C** [comment]
I've been trying to reproduce this by re-running the CI on master https://github.com/bitcoin/bitcoin/commit/49b87bfe7e2799d25ce709123ecafa872b36e87a: https://cirrus-ci.com/task/4749392670883840

So far all runs passed..

**Comment by 0xB10C** [comment]
Thanks, didn't know `rr`. Especially the chaos mode (https://robert.ocallahan.org/2016/02/introducing-rr-chaos-mode.html) seems to be helpful here. However,  rr and bpf don't work well together yet. `rr` asserts when running the python USDT interface tests. I've tried to patch `rr` to make this work but haven't been successful yet.

**Comment by 0xB10C** [comment]
Gave this another look. No luck with `rr` and haven't been able to reproduce this locally or in the CI.

**Comment by mzumsande** [comment]
> Maybe it will "fix itself" after #27360 ?

Unfortunately, it didn't (https://github.com/bitcoin/bitcoin/runs/13206824577 from master after #27360 was merged)

**Comment by fanquake** [comment]
master @ e0a70c5b4f2c691e8d6b507c8ce879f0b0424254. https://cirrus-ci.com/task/5247512164958208:
```bash
 node0 2023-05-10T17:29:06.449483Z (mocktime: 2023-05-24T17:29:09Z) [httpworker.1] [rpc/request.cpp:181] [parse] [rpc] ThreadRPCServer method=getrawmempool user=__cookie__ 
 test  2023-05-10T17:29:06.450000Z TestFramework (INFO): Polling buffer... 
 test  2023-05-10T17:29:06.486000Z TestFramework (INFO): Ensuring mempool:rejected event was handled successfully... 
 test  2023-05-10T17:29:06.486000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 339, in run_test
                                       self.rejected_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 323, in rejected_test
                                       assert_equal(EXPECTED_REJECTED_EVENTS, handled_rejected_events)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 55, in assert_equal
                                       raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
                                   AssertionError: not(1 == 0)
 test  2023-05-10T17:29:06.487000Z TestFramework (DEBUG): Closing down network thread
...
interface_usdt_mempool.py                              | ✖ Failed  | 5 s
```

**Comment by 0xB10C** [comment]
I'll give this another look next week.

**Comment by virtu** [comment]
@0xB10C and I had a look at this today. Here's what we found so far:
1. We believe the issue is caused by [one or more of the assertions](https://github.com/bitcoin/bitcoin/blob/b22408df162a224d94ac54e8443b57ef3fd2ca72/test/functional/interface_usdt_mempool.py#L306-L307) failing in the `handle_rejected_event` callback (although we are not sure why they fail).
2. The test framework's `assert_equal()` communicates failed assertions via exceptions. Unfortunately exceptions raised in bcc callback functions are not properly propagated (`Exception ignored on calling ctypes callback function: <function PerfEventArray._open_perf_buffer.<locals>.raw_cb_ at 0xffff83490e50>`), so they only appear on stderr, not in the logs. As a consequence, the logs only show [that the test case failed](https://github.com/bitcoin/bitcoin/blob/b22408df162a224d94ac54e8443b57ef3fd2ca72/test/functional/interface_usdt_mempool.py#L326), not why it failed.
3. Because (2) is a general problem with functional tests for tracepoints, I'll create a PR that fixes the issue for all of them. The best solution we came up with so far is wrapping all `assert_equals()` in bcc callback functions in `try` blocks and logging any `AssertionError` exceptions using `self.log()`. Once this (or a better solution, preferable one that involves not ignoring exceptions in ctype callback functions, if that's possible) is merged, the logs should tell us why the `mempool:rejected` test fails sporadically.

**Comment by virtu** [comment]
Since the failure hasn't been observed since https://github.com/bitcoin/bitcoin/pull/27177 was merged, there's a good chance it fixed the issue.

However, it still seems like a good idea to address the issue of missing `check_equal()` exception information in the logs, so I created https://github.com/bitcoin/bitcoin/pull/27831 to make sure all tracepoint tests properly log failed `check_equal()` assertions.

**Comment by 0xB10C** [comment]
> Can it be forced to "reproduce" on master by adding a typo to the error string in assert_equal?

see https://github.com/0xB10C/bitcoin/commit/4091d621f16b6316eef5551a56c00bf0e2c0e541 and this CI task https://cirrus-ci.com/task/4909169740873728

**Comment by glozow** [comment]
https://cirrus-ci.com/task/6053265947754496 ? 😢

**Comment by fanquake** [comment]
https://cirrus-ci.com/task/5268139265490944

**Comment by fanquake** [comment]
https://cirrus-ci.com/task/6504094303518720?logs=ci#L3264
```bash
 test  2023-07-03T14:16:38.121000Z TestFramework (INFO): Polling buffer... 
 node0 2023-07-03T14:16:38.121067Z (mocktime: 2023-07-17T14:16:35Z) [http] [httpserver.cpp:255] [http_request_cb] [http] Received a POST request for / from 127.0.0.1:40312 
 node0 2023-07-03T14:16:38.121207Z (mocktime: 2023-07-17T14:16:35Z) [httpworker.2] [rpc/request.cpp:181] [parse] [rpc] ThreadRPCServer method=getrawmempool user=__cookie__ 
 test  2023-07-03T14:16:38.122000Z TestFramework (INFO): Ensuring mempool:rejected event was handled successfully... 
 test  2023-07-03T14:16:38.122000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 131, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 329, in run_test
                                       self.rejected_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_mempool.py", line 310, in rejected_test
                                       assert_equal(event.reason.decode("UTF-8"), "min relay fee not met")
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/util.py", line 57, in assert_equal
                                       raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
                                   AssertionError: not( == min relay fee not met)
 test  2023-07-03T14:16:38.123000Z TestFramework (DEBUG): Closing down network thread 
 test  2023-07-03T14:16:38.173000Z TestFramework (INFO): Stopping nodes 
 test  2023-07-03T14:16:38.173000Z TestFramework.node0 (DEBUG): Stopping node 
```

**Comment by virtu** [comment]
Even with the extra information, I'm somewhat at a loss because I cannot reproduce the problem locally. From what I understand:
- the tracepoint got called, the handler was executed
- the tx hash is correct (so `bpf_usdt_readarg_p(1, ctx, &rejected.hash, HASH_LENGTH)` in `trace_rejected` works as expected)
- reject reason is usually correct; but sometime's it's an empty string, suggesting something went wrong with `bpf_usdt_readarg_p(2, ctx, &rejected.reason, MAX_REJECT_REASON_LENGTH)` in `trace_rejected`. Why this only would only happen somethings, I do not know. If the environment is 100% reproducible, this should not happen.

My first reaction to your suggestion of removing the sub-test case was to first try and change the code so it triggers another rejection reason. But without knowing what's going wrong right now, this feels like poking around in the dark.

**Comment by willcl-ark** [comment]
AFAIU the macros themselves don't copy any data, rather just define a static probe point. Any arguments passed to them just get incorporated into the assembly code. The tracing program (BPF) if loaded into the kernel can run code at the trace points. Here it can choose to, and AFAIU does in our case with BPF, copy the data into a different kernel space map when it hits a tracepoint, which can then be read by the USDT probe in our python test.

So the data does get copied but not by the macro, rather by the BPF program.

**Comment by willcl-ark** [comment]
I'm probably outside of my wheelhouse now, but as I understand the tracepoint defines a place in the binary where a tracing program gets to insert and run (its own) code, if it's loaded in the kernel. The arguments to the macro are passed directly to the (assembly) code provided by the tracing program _in that expression_ and at that execution point BPF will copy values into it's own map in the kernel if a probe is set up. Other tracing programs or runtimes can choose not to copy data if they don't need/want to.

By using helper function `bpf_usdt_readarg` in our test program we should be able to read (pointers) between kernel and user space safely too. In the recent failure here, in theory, the macro _should_ handle the pointers ok as neither `tx` nor `result` go out of scope before it's called, at which point the copy should be done.

All that said, it's obviously still going wrong somehow...

**Comment by fanquake** [comment]
I guess we can close this for now, and can reopen if #30559 is reverted?

**Comment by 0xB10C** [comment]
I'm once again trying to (reliably) reproduce this to be able to have a chance at debugging it. My current approach is running the mempool:rejected test 100 times along with the other functional tests (https://github.com/0xB10C/bitcoin/tree/2024-08-27380-intermitting-usdt-mempool-failures). 

However, no luck yet getting it to a point where I can reproduce it reliably. Sometimes [all 100 runs succeed](https://github.com/0xB10C/bitcoin/actions/runs/10255800979/job/28373512626), other times it fails at the [second run](https://github.com/0xB10C/bitcoin/actions/runs/10255800979/job/28375206006#step:7:4704). Currently [re-running it one more time](https://github.com/0xB10C/bitcoin/actions/runs/10255800979/job/28375569586).

**Comment by 0xB10C** [comment]
Thanks! Did a few CI run of https://github.com/0xB10C/bitcoin/commit/3d806eafa04415791b95ed20525eed2fd54f5b38. Your patch doesn't seem to improve the situation.

- Attempt 1 failed: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/1 (test iteration 4/100)
- Attempt 2 failed: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/2 (test iteration 18/100)
- Attempt 3 succeded: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/3
- Attempt 4 succeded: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/4
- Attempt 5 failed: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/5 (test iteration 26/100)
- Attempt 6 succeded: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/6
- Attempt 7 failed: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/7 (test iteration 11/100)
- Attempt 8 succeded: https://github.com/0xB10C/bitcoin/actions/runs/10348240621/attempts/8

**Comment by mzumsande** [comment]
Maybe copying the string into a static or global variable would help rule out a lifetime issue?

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 9126d8e928..683d13e0e5 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1863,6 +1863,8 @@ MempoolAcceptResult AcceptToMemoryPool(Chainstate& active_chainstate, const CTra
     CTxMemPool& pool{*active_chainstate.GetMempool()};
 
     std::vector<COutPoint> coins_to_uncache;
+    static char string_buffer[120];
+
     auto args = MemPoolAccept::ATMPArgs::SingleAccept(chainparams, accept_time, bypass_limits, coins_to_uncache, test_accept);
     MempoolAcceptResult result = MemPoolAccept(pool, active_chainstate).AcceptSingleTransaction(tx, args);
     if (result.m_result_type != MempoolAcceptResult::ResultType::VALID) {
@@ -1873,9 +1875,10 @@ MempoolAcceptResult AcceptToMemoryPool(Chainstate& active_chainstate, const CTra
 
         for (const COutPoint& hashTx : coins_to_uncache)
             active_chainstate.CoinsTip().Uncache(hashTx);
+        strncpy(string_buffer, result.m_state.GetRejectReason().c_str(), 120 - 1);
         TRACE2(mempool, rejected,
                 tx->GetHash().data(),
-                result.m_state.GetRejectReason().c_str()
+                string_buffer
         );
     }
     // After we've (potentially) uncached entries, ensure our coins cache is still within its size limits

**Comment by 0xB10C** [comment]
Thanks @mzumsande - [all 11 CI runs with your patch succeed](https://github.com/0xB10C/bitcoin/actions/runs/10359982384/job/28712303493). 

This should confirm it's indeed a lifetime issue.

**Comment by mzumsande** [comment]
This is still really confusing to me.

I don't understand tracepoints in depth, but it looks like there is a race between bicoind continuing to run (and overwriting data after it goes out of scope), and the data-collecting program (bpftrace) reading data from the memory location  the `c_str()` result points to. But if that was the case, it would be a very general problem with tracepoints, and I don't see how it would specifically lead to failures in this exact situation with the `mempool-rejected` tracepoint.

I also don't understand why this specifically happens in CI runs with sanitizers enabled. Naively I would expect that sanitizers make the C++ program run slower (while not affecting the speed of bpftrace), so if there really was a race, I would expect failures to happen less often with sanitizers, not more.

**Comment by 0xB10C** [comment]
To fix a potential lifetime issue, it is probably helpful to replace `bpf_usdt_readarg_p()` (which isn't even documented) in:

https://github.com/bitcoin/bitcoin/blob/f7144b24be09e7db2173a0b15d73f99a08b98118/test/functional/interface_usdt_mempool.py#L103-L104

with `bpf_usdt_readarg()` followed by [`bpf_probe_read_user()`](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#10-bpf_probe_read_user) or a [`bpf_probe_read_user_str()`](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#11-bpf_probe_read_user_str). We use `bpf_usdt_readarg_p()` in multiple places.

From the documentation:
> For safety, all user address space memory reads must pass through bpf_probe_read_user().


This is also how it's documented in ["6. USDT probes - bcc reference guide"](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#6-usdt-probes) and how it's done in e.g. this example [`mysqld_query.py`](https://github.com/iovisor/bcc/blob/8d85dcfac86bb7402a20bea5ceba373e5e019b6c/examples/tracing/mysqld_query.py#L36-L37).

Happy to give this another spin.

**Comment by 0xB10C** [comment]
> Happy to give this another spin.

To check that this is still an issue, I ran https://github.com/bitcoin/bitcoin/compare/master...0xB10C:bitcoin:refs/heads/2024-11-27380-intermitting-usdt-mempool-failures in CI [20 times](https://github.com/0xB10C/bitcoin/actions/runs/12040804871). It failed 6 times. I then ran it with https://github.com/0xB10C/bitcoin/commit/8a3eeb48482955e9d0cd26632a03457e35a10568 and it didn't fail a single time. It seems that getting rid of `bpf_usdt_readarg_p` indeed fixes this.

I'll work on a PR to get rid of our `bpf_usdt_readarg_p` usage in all tests.

**Comment by 0xB10C** [comment]
opened #31848 which reverts https://github.com/bitcoin/bitcoin/pull/30574 and should fix the intermittent failures.
